### PR TITLE
Make pylabrad work with python3

### PR DESCRIPTION
**Pull request by maffoo**
Turns out that twisted now works with python 3 well enough for us to use, so we really have no excuse not to go for it. It turned out to be not to much work to get everything working. I tried to keep the commits mostly small on this branch, though the last one in particular is a bit of a grab bag. I'm using the ["future" package](http://python-future.org/index.html) for compatibility, with the goal being to use `from __future__` imports and imports from the `future` module to write code that is python 3-compatible as far as possible, but is also backwards-compatible with python 2.

str/bytes handling is a bit awkward, but not too bad for the most part since we added a byte type "y" to labrad. It's also a bit tricky to deal with labrad signed ints ("i") and unsigned ints ("w") since we previously used python2's `int` and `long` types to correspond to these; python3 no longer has `long` but in most cases this doesn't matter because we usually don't care in python code, and we can use type inference when sending to labrad servers.

### CODE REVIEW

File: labrad/node/__init__.py [hunk 1]
```diff
@@ -517,8 +517,8 @@ def _packet(self):
     def _update(self, config, triggerRefresh=True):
         """Update instance variables from loaded config."""
         self.dirs, self.extensions, self.autostart = config
-        print 'config updated: dirs={}, extensions={}, autostart={}'.format(
-                self.dirs, self.extensions, self.autostart)
+        print('config updated: dirs={}, extensions={}, autostart={}'.format(
+        self.dirs, self.extensions, self.autostart))
```
Comment (kunalq) [inline labrad/node/__init__.py]: Missing whitespace :)

File: labrad/pyparsing.py [hunk 2]
```diff
@@ -562,13 +562,13 @@ def line(loc, strg):
         return strg[lastCR+1:]
 
 def _defaultStartDebugAction(instring, loc, expr):
-    print "Match", _ustr(expr), "at loc", loc, "(%d,%d)" % (lineno(loc, instring), col(loc, instring))
+    print("Match", _ustr(expr), "at loc", loc, "(%d,%d)" % (lineno(loc, instring), col(loc, instring)))
```
Comment (kunalq) [inline labrad/pyparsing.py]: Should probably be switched to `.format`

File: labrad/pyparsing.py [hunk 3]
```diff
@@ -562,13 +562,13 @@ def line(loc, strg):
...
 
 def _defaultSuccessDebugAction(instring, startloc, endloc, expr, toks):
-    print "Matched", _ustr(expr), "->", toks.asList()
+    print("Matched", _ustr(expr), "->", toks.asList())
 
 def _defaultExceptionDebugAction(instring, loc, expr, exc):
-    print "Exception raised:", _ustr(exc)
+    print("Exception raised:", _ustr(exc))
```
Comment (kunalq) [inline labrad/pyparsing.py]: Ditto

File: labrad/pyparsing.py [hunk 4]
```diff
@@ -689,7 +689,7 @@ def _normalizeParseActionArgs(f):
                 if hasattr(f.__call__, "im_self"):
                     numargs -= 1
 
-        #~ print "adding function %s with %d args" % (f.func_name,numargs)
+        #~ print("adding function %s with %d args" % (f.func_name,numargs))
```
Comment (kunalq) [inline labrad/pyparsing.py]: nit -- remove commented code :)

File: labrad/pyparsing.py [hunk 5]
```diff
@@ -796,7 +796,7 @@ def _parseNoCache(self, instring, loc, doActions=True, callPreParse=True):
         debugging = self.debug #and doActions
 
         if debugging or self.failAction:
-            #~ print "Match", self, "at loc", loc, "(%d,%d)" % (lineno(loc, instring), col(loc, instring))
+            #~ print("Match", self, "at loc", loc, "(%d,%d)" % (lineno(loc, instring), col(loc, instring)))
```
Comment (kunalq) [inline labrad/pyparsing.py]: Ditto :)

File: labrad/pyparsing.py [hunk 6]
```diff
@@ -2922,7 +2922,7 @@ def oneOf(strs, caseless=False, useRegex=True):
             i += 1
 
     if not caseless and useRegex:
-        #~ print strs,"->", "|".join( [ _escapeRegexChars(sym) for sym in symbols] )
+        #~ print(strs,"->", "|".join( [ _escapeRegexChars(sym) for sym in symbols] ))
```
Comment (kunalq) [inline labrad/pyparsing.py]: And a few more...

File: labrad/servers/test_server.py [hunk 7]
```diff
@@ -193,7 +194,7 @@ def set(self, c, key, value):
 
     @setting(101, "Get", key='s', returns='?')
     def get(self, c, key):
-        print "getting tag: %s, value: %s" % (key, c['dict'][key])
+        print("getting tag: %s, value: %s" % (key, c['dict'][key]))
```
Comment (kunalq) [inline labrad/servers/test_server.py]: Switch to `.format`

File: labrad/servers/threaded_test_server.py [hunk 8]
```diff
@@ -145,7 +147,7 @@ def set(self, c, key, value):
 
     @setting(101, "Get", key='s', returns='?')
     def get(self, c, key):
-        print "getting tag: %s, value: %s" % (key, c['dict'][key])
+        print("getting tag: %s, value: %s" % (key, c['dict'][key]))
```
Comment (kunalq) [inline labrad/servers/threaded_test_server.py]: Switch to `.format`

File: labrad/test/unit_performance.py [hunk 9]
```diff
@@ -132,13 +135,13 @@ def main():
     result = run_tests()
 
     if args.format == 'json':
-        print json.dumps({k: v['ms'] for k, v in result.items()}, indent=2)
+        print(json.dumps({k: v['ms'] for k, v in result.items()}, indent=2))
```
Comment (kunalq) [inline labrad/test/unit_performance.py]: This is a side note, but this is a good candidate for switching to constants `ms` instead of `'ms'`

File: labrad/util/__init__.py [hunk 10]
```diff
@@ -343,8 +345,8 @@ class ServerOptions(usage.Options):
             tls_on = config['tls'] == 'on'
             config['port'] = C.MANAGER_PORT_TLS if tls_on else C.MANAGER_PORT
     except usage.UsageError, errortext:
-        print '%s: %s' % (sys.argv[0], errortext)
-        print '%s: Try --help for usage details.' % (sys.argv[0])
+        print('%s: %s' % (sys.argv[0], errortext))
+        print('%s: Try --help for usage details.' % (sys.argv[0]))
```
Comment (kunalq) [inline labrad/util/__init__.py]: Switch to `.format`

File: labrad/wrappers.py [hunk 11]
```diff
@@ -542,8 +543,8 @@ def _refreshServer(self, name):
         try:
             yield server.refresh()
         except Exception, e:
-            print 'Error while refreshing server "%s":' % name
-            print repr(e)
+            print('Error while refreshing server "%s":' % name)
+            print(repr(e))
```
Comment (kunalq) [inline labrad/wrappers.py]: These four could also benefit from a `.format` change :)

File: labrad/pyparsing.py [hunk 12]
```diff
@@ -1425,7 +1425,7 @@ def __init__(self, initChars, bodyChars=None, min=1, max=0, exact=0, asKeyword=F
         self.maxSpecified = max > 0
 
         if min < 1:
-            raise ValueError, "cannot specify a minimum length < 1; use Optional(Word()) if zero-length word is permitted"
+            raise ValueError("cannot specify a minimum length < 1; use Optional(Word()) if zero-length word is permitted")
```
Comment (kunalq) [inline labrad/pyparsing.py]: Good opportunity to break this line up (including ones below)

File: labrad/decorators.py [hunk 13]
```diff
@@ -28,6 +28,7 @@
 import types
 
 import twisted.internet.defer as defer
+from past.builtins import basestring  # for python 2/3 compatibility
```
Comment (kunalq) [inline labrad/decorators.py]: I tried importing this in 2.7 -- it looks like it's third party. Is it a part of the `future` library?

File: labrad/types.py [hunk 14]
```diff
@@ -447,16 +453,13 @@ def reprLRData(s):
 
 # LabRAD type classes
 
-class LRType(object):
+class LRType(with_metaclass(RegisterParser, object)):
```
Comment (kunalq) [inline labrad/types.py]: Interesting -- the `six` package also contains this shim. Is there a difference between `six`'s implementation and `future`'s?

File: labrad/types.py [hunk 15]
```diff
@@ -707,13 +725,13 @@ def __unflatten__(self, s, endianness):
...
+        if not isinstance(s, bytes):
             raise FlatteningError(s, self)
         return struct.pack(endianness + 'I', len(s)) + s, self
 
-# Since bytes is an alias for str in python 2.7, don't enable this
-# until we are ready to cut over.
-# registerType(bytes, LRBytes)
+# bytes is an alias for str in python 2.7, but in python 3 they are different.
+if bytes != str:
+    registerType(bytes, LRBytes())
```
Comment (kunalq) [inline labrad/types.py]: I'd suggest being more explicit here -- e.g.
```python
if (sys.version_info > (3, 0)): ...
```

File: labrad/types.py [hunk 16]
```diff
@@ -1319,8 +1337,8 @@ def __flatten_array__(self, a, endianness):
         else:
             elems = (self.elem.flatten(i, endianness=endianness).bytes
                      for i in a.flat)
-            return dims + ''.join(elems), self
-        return dims + a.tostring(), self
+            return dims + b''.join(elems), self
+        return dims + a.tobytes(), self
```
Comment (kunalq) [inline labrad/types.py]: In Python 2.7, won't this throw an exception? Or does the `future` library monkey patch strings?

File: labrad/node/__init__.py (hunk 1 above)
Comment (maffoo) [inline labrad/node/__init__.py]: fixed

File: labrad/servers/test_server.py (hunk 7 above)
Comment (maffoo) [inline labrad/servers/test_server.py]: done.

File: labrad/servers/threaded_test_server.py (hunk 8 above)
Comment (maffoo) [inline labrad/servers/threaded_test_server.py]: done.

File: labrad/types.py (hunk 14 above)
Comment (maffoo) [inline labrad/types.py]: This version is slightly different from the one in `six`; according to the docstring:
> This has the advantage over six.with_metaclass of not introducing dummy classes into the final MRO.

File: labrad/types.py (hunk 15 above)
Comment (maffoo) [inline labrad/types.py]: The official [porting guide](https://docs.python.org/3/howto/pyporting.html#use-feature-detection-instead-of-version-detection) suggests using feature detection rather than version detection, which is why I did it this way. I'm open to changing it though if you feel like this is unclear.

File: labrad/types.py (hunk 16 above)
Comment (maffoo) [inline labrad/types.py]: `a` is a numpy array, not a string, and numpy arrays have a [`tobytes` method](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.tobytes.html#numpy.ndarray.tobytes) since version 1.9.0 (we're requiring numpy 1.9.1 or later, so this should be ok).

File: labrad/test/unit_performance.py (hunk 9 above)
Comment (maffoo) [inline labrad/test/unit_performance.py]: I'll file a separate issue about using unit objects instead of strings.

Comment (maffoo) [inline labrad/test/unit_performance.py]: #309

**Comment by jayich** [comment]
@maffoo , thank you!

**Comment by maffoo** [comment]
@jayich, glad you're excited to try this. I'm sure there are some gaps because our test coverage is not 100% (currently working on some additional fixes for the node server, for example), but please feel free to test with python3 and report any issues you have. We'll try to get things ironed out before we make an official release for python3.

**Comment by DanielSank** [comment]
I don't see any obvious issues. Thanks much for the changes, @maffoo.

**Comment by jayich** [comment]
@maffoo , any updates on the official Python 3 release?  I am tempted to dive into Python 3 with pylabrad in a fresh environment.  Let me rephrase: are there any updates from Python 3 users with pylabrad?  I am happy to be the first one, but would feel more comfortable knowing that I am not the only one (yet).  Is your group in the process of switching everything to Python 3?

Also, I imagine if I am stuck with some module that only works on Python 2 (say proprietary) I could run that on an isolated computer and it would probably connect fine to the rest of the labrad network (or maybe not?).

**Comment by maffoo** [comment]
@jayich, yes, the network protocol is identical, so you should have no problem mixing and matching labrad servers and clients running on python2 and python3. I haven't heard from folks running on python3 yet, but we have hammered out some additional compatibility issues such as getting the node server to work, so I'm reasonably confident that things will go smoothly. We aren't currently pushing on moving to python3 because we have a large body of python2-only code, but at least now we can't use pylabrad itself as our excuse :-)

**Comment by jayich** [comment]
@maffoo , great.  The ability to mix and match (thank you labrad and especially the developers) makes the choice to use Python 3 easy.
